### PR TITLE
[FIRRTL] Donot run MemToReg transformation on Testbench Memories

### PR DESCRIPTION
**Pull request by prithayan**
Run the `MemToRegofVec` transformation only on `DUT` memories and ignore any modules in the testbench. 
Preserve all other MemOps, to be handled by either `LowerMemory` for lowering to 
extern modules for macro replacement, or by `HWMemSimImpl` for lowering to registers.

### CODE REVIEW

**Comment by youngar** [comment]
How is `MemToReg` memory lowering different than `HWMemSimImpl`?  Do we need both? Sometime soon, while it's still paged in, we should update the FIRRTLRationale to cover all the steps involved in memory lowering.

LowerMemory complains about any non-seq mem it finds and prints an error, but it should be easy to change `emitError` to `continue`.

**Comment by prithayan** [comment]
@youngar , I think the difference in `LowerTypes`. 
I missed the important comment in the scala code: 
`@note Only applies to the MarkedDUT`
`federation/blob/master/src/main/scala/firrtl/MemToRegOfVecTransform.scala#L23`. 
Updated the pass to work on only on DUT modules. Can you please check it once.

File: lib/Dialect/FIRRTL/Transforms/MemToRegOfVec.cpp [hunk 1]
```diff
@@ -29,10 +32,38 @@ namespace {
...
+    });
+    if (it != body->end()) {
+      auto &instanceGraph = getAnalysis<InstanceGraph>();
+      auto *node = instanceGraph.lookup(&(*it));
+      llvm::for_each(llvm::depth_first(node), [&](hw::InstanceGraphNode *node) {
+        dutModuleSet.insert(node->getModule());
+      });
+    }
+
+    std::vector<FModuleLike> ops;
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/MemToRegOfVec.cpp:56]:
```suggestion
```

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp [hunk 2]
```diff
@@ -474,7 +474,7 @@ LogicalResult LowerMemoryPass::runOnModule(FModuleOp module, bool shouldDedup) {
     if (!((summary.readLatency == 1 && summary.writeLatency == 1) &&
           (summary.numWritePorts + summary.numReadWritePorts == 1) &&
           (summary.numReadPorts <= 1) && summary.dataWidth > 0))
-      return op->emitError("only seqmems are supported by LowerMemory");
+      continue;
```
Comment (youngar) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:477]: have to update the test for this as well

File: test/Dialect/FIRRTL/SFCTests/data-taps.fir [hunk 3]
```diff
@@ -3,7 +3,10 @@
 ; - github.com/sifive/$internal:
 ;   - src/test/scala/grandcentral/DataTapsTest.scala
 
-circuit Top :
+circuit Top :%[[{
+   "class":"sifive.enterprise.firrtl.MarkDUTAnnotation" ,
+    "target":"~Top|Child"
+     }]]
```
Comment (richardxia) [inline test/Dialect/FIRRTL/SFCTests/data-taps.fir:9]: Nit: The whitespace and alignment are slightly bothering me. If this is legal, can we format it like this?

```suggestion
circuit Top : %[[{
    "class": "sifive.enterprise.firrtl.MarkDUTAnnotation",
    "target": "~Top|Child"
  }]]
```

I can be pretty flexible, but mostly just want the keys to line up :)

Same comment applies to several of the other tests where you had to add these annotations.

**Review by youngar — APPROVED** [review]

File: lib/Dialect/FIRRTL/Transforms/MemToRegOfVec.cpp [hunk 4]
```diff
@@ -26,14 +29,45 @@ using namespace circt;
...
+    auto circtOp = getOperation();
+    static const char dutAnnoClass[] =
+        "sifive.enterprise.firrtl.MarkDUTAnnotation";
+    DenseSet<Operation *> dutModuleSet;
+
+    auto *body = circtOp.getBody();
+
+    // Find the device under test and create a set of all modules underneath it.
+    auto it = llvm::find_if(*body, [&](Operation &op) -> bool {
+      return AnnotationSet(&op).hasAnnotation(dutAnnoClass);
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/MemToRegOfVec.cpp:48]: For future reference, we should move DUT determination into an analysis so passes don't have to be CircuitOp passes.

**Pull request merged**